src/main/java/com/commercetools/sync/commons/utils/CompletableFutureUtils.java

```diff
@@ -0,0 +1,162 @@
+                                                                     .collect(Collectors.toSet()));
+    }
+
+
+    /**
+     * Maps a stream of values to futures using the supplied mapper function.
+     *
+     * @param values stream of values to apply a mapper function that would map each to a {@link CompletionStage}.
+     * @param mapper function to map each value to a {@link CompletionStage}
+     * @param <T>    The type of the values.
+     * @param <S>    The type of the mapped values.
+     * @return a list of futures resulting from applying the mapper function on each value.
+     */
+    @Nonnull
+    public static <T, S> List<CompletableFuture<S>> mapValuesToFutureList(
+        @Nonnull final Stream<T> values,
+        @Nonnull final Function<T, CompletionStage<S>> mapper) {
+        return toListOfCompletableFutures(map(values, mapper));
+    }
+
+    /**
+     * Maps a stream of values to futures using the supplied mapper function.
+     *
+     * @param values stream of values to apply a mapper function that would map each to a {@link CompletionStage}.
+     * @param mapper function to map each value to a {@link CompletionStage}
+     * @param <T>    The type of the values.
+     * @param <S>    The type of the mapped values.
+     * @return a list of futures resulting from applying the mapper function on each value.
+     */
+    @Nonnull
+    public static <T, S> Set<CompletableFuture<S>> mapValueSetToFutureSet(
+        @Nonnull final Set<T> values,
+        @Nonnull final Function<T, CompletionStage<S>> mapper) {
+        return toSetOfCompletableFutures(map(values.stream(), mapper));
+    }
+
+
+    /**
+     * Converts a stream of {@link CompletionStage} of values of type {@code <T>} into a
+     * {@link List}{@link CompletableFuture} of values of type {@code <T>}.
+     *
+     * @param values stream of {@link CompletionStage} of values of type {@code <T>}
+     * @param <T>    the type of the results of the stages.
+     * @return a {@link List} of {@link CompletableFuture} elements of type {@code <T>}.
+     */
+    @Nonnull
+    public static <T> List<CompletableFuture<T>> toListOfCompletableFutures(
+        @Nonnull final Stream<CompletionStage<T>> values) {
+        return values.map(CompletionStage::toCompletableFuture).collect(toList());
+    }
+
+    /**
+     * Converts a stream of {@link CompletionStage}&lt;T&gt; into a
+     * {@link Set}&lt;{@link CompletableFuture}&lt;T&gt;&gt;.
+     *
+     * @param values stream of {@link CompletionStage} of type {@code <T>}
+     * @param <T>    the type of the results of the stages.
+     * @return a {@link Set} of {@link CompletableFuture}&lt;T&gt;.
+     */
+    @Nonnull
+    public static <T> Set<CompletableFuture<T>> toSetOfCompletableFutures(
+        @Nonnull final Stream<CompletionStage<T>> values) {
+        return values.map(CompletionStage::toCompletableFuture).collect(toSet());
+    }
+
+    private CompletableFutureUtils() {
+    }
+}
```

Comment:
It looks like the utility could be re-factored:
  - unified methods to accept `Collections`, instead of `List` and `Set` separately, and then use `Collections#stream()`
    Maybe it should be overloaded `mapValuesToFutureOfCompletedValues(Collection<T> values, ...)`
  - `toListOfCompletableFutures` and `toSetOfCompletableFutures` could be unified using second argument of type `Collector`, and based on this collector type it will return `List` or `Set`

Then in most cases calls like `entities.entrySet().stream()` could be encapsulated passing just `entities`

Reply:
Utils are refactored: 
6d404c9

Reply:
since you use `mapValuesToFutureOfCompletedValues()` mostly with `toList`, does it have sense to create a dedicated function <code>mapValuesToFutureOf<b>List</b>CompletedValues</code>

Reply:
smth like:
```java
@Nonnull
public static <T, S> CompletableFuture<List<S>> mapValuesToFutureOfListCompletedValues(
    @Nonnull final Collection<T> values,
    @Nonnull final Function<T, CompletionStage<S>> mapper) {
    return mapValuesToFutureOfCompletedValues(values.stream(), mapper, toList());
}

@Nonnull
public static <T, S> CompletableFuture<Set<S>> mapValuesToFutureOfSetCompletedValues(
    @Nonnull final Collection<T> values,
    @Nonnull final Function<T, CompletionStage<S>> mapper) {
    return mapValuesToFutureOfCompletedValues(values.stream(), mapper, toSet());
}
```

Reply:
@andrii-kovalenko-ct would be good, but there I don't have the time luxury to do that now (since I will add more tests and so on for it). There are more important things to do now. So maybe later: https://github.com/commercetools/commercetools-sync-java/issues/258

Reply:
ok